[native] Add CTAS tests

pdabre12 opened the pull request:
Resolves: https://github.com/prestodb/presto/issues/20848

pdabre12 commented:
@majetideepak Please review this PR.

pramodsatya commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeCtasQueries.java:
```diff
@@ -0,0 +1,152 @@
+        extends AbstractTestQueryFramework {
+
+    @Override
+    protected void createTables() {
```
Why are all the tables created here? Only the tables nation, region and orders seem to be used in the tests.

pramodsatya commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeCtasQueries.java:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    @Test
+    public void testCreateTableAsSelect() {
```
This appears identical to the function `testCreateTableAsSelect` in `presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java`. It could be reused by adding another abstract class for CTAS tests.

pdabre12 replied:
I am adding some more queries later on, so need to override it anyways.

pramodsatya commented on presto-native-execution/pom.xml:
```diff
@@ -156,6 +156,11 @@
             <groupId>com.facebook.airlift</groupId>
             <artifactId>log-manager</artifactId>
         </dependency>
+        <dependency>
```
This can be reverted after the test utility functions are moved to `AbstractTestQueryFramework.java`.

pdabre12 replied:
Because the test utility function was overridden , had to keep this change.

pramodsatya reviewed (commented):
Thanks for adding the tests @pdabre12, did an initial parse and left some comments.

majetideepak commented:
We want Parquet as the file format as well.

czentgr commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeCtasQueries.java:
```diff
@@ -0,0 +1,152 @@
+    protected void assertCreateTableAsSelect(Session session, String table, @Language("SQL") String query, @Language("SQL") String expectedQuery, @Language("SQL") String rowCountQuery) {
+        assertUpdate(session, "CREATE TABLE " + table + " AS " + query, rowCountQuery);
+        assertQuery(session, "SELECT * FROM " + table, expectedQuery);
+        assertUpdate(session, "DROP TABLE " + table);
```
Do you need to test DROP TABLE in the CTAS test?
You would need to allow DROP TABLE in https://github.com/prestodb/presto/blob/dd5b168ff5bdf306de70be19cc37c9797ca00981/presto-native-execution/src/test/java/com/facebook/presto/nativeworker/NativeQueryRunnerUtils.java

and add `hive.allow-drop-table=true` as a hive property so you can drop the table - after you found that this is missing.

pdabre12 replied:
Thanks @czentgr for the comments. I did exactly that and it works , but these tests will only work locally/on my branch unless `hive.allow-drop-table=true` is being set in the main codebase. Should I add the property `hive.allow-drop-table=true` in the codebase and try to raise a different PR mentioning the access issues?

czentgr replied:
That's why I'm asking if you truly need to test DROP table here. After all, this is a pure catalog operation which certainly is tested and possibly doesn;t need to be done for the native worker as it is unrelated.

pdabre12 replied:
It's not needed per say. Part of the reason why I was using them was because I don't want many tables to get created and persist there. I still do not know whether the tables are deleted after the tests are ran or they persist. If they are deleted, DROP operation is not needed.

czentgr replied:
Yes, normally a good though to clean up. In this case, afaik, it is not needed because they won't persist past the test run in the hive metastore. Note how new tables are created at the beginning of the test as well.

pdabre12 replied:
Okay, understood.

majetideepak replied:
I think we should stick with the DROP operation always. I agree with @pdabre12 that cleaning up is ideal. Enabling the API to not delete the tables can lead to bugs.
I don't see any harm in adding `hive.allow-drop-table=true` to the Native runner. I assume the Java runner has this as well.

pdabre12 commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeCtasQueries.java:
```diff
@@ -0,0 +1,152 @@
+        assertUpdate("DROP TABLE analyze_test");
+
+    }
+    protected void assertCreateTableAsSelect(String table, @Language("SQL") String query, @Language("SQL") String rowCountQuery)
```
I did that , but now I am not using the DROP operation , hence need to override the method.

pdabre12 commented:
@pramodsatya Made some changes based on your comments and added some files, some of the test cases are failing. 
I am creating an excel sheet showcasing all the passing/failing operations.

pramodsatya commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/TestPrestoNativeCtasQueriesParquetUsingJSON.java:
```diff
@@ -0,0 +1,36 @@
+import com.facebook.presto.testing.QueryRunner;
+import org.testng.annotations.Test;
+
+@Test(groups = {"parquet"})
```
Testing with thrift format is enough and the test files using Json format can be removed.

pramodsatya commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeCtasQueries.java:
```diff
@@ -0,0 +1,127 @@
+    {
+        assertUpdate(session, "CREATE TABLE " + table + " AS " + query, rowCountQuery);
+        assertQuery(session, "SELECT * FROM " + table, expectedQuery);
+        assertFalse(getQueryRunner().tableExists(session, table));
```
This should be reverted since the table is not being dropped. This function can also be moved to AbstractTestQueryFramework so the dependency on intellij annotations can be removed.

pramodsatya reviewed (commented):
Thanks for making the changes @pdabre12, left some comments. Could you please look into the test failures and ensure they are passing? The changes to submodule velox also need to be reverted.

pdabre12 commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/TestPrestoNativeCtasQueriesDwrfUsingThrift.java:
```diff
@@ -0,0 +1,34 @@
+    protected QueryRunner createQueryRunner()
+            throws Exception
+    {
+        return PrestoNativeQueryRunnerUtils.createNativeQueryRunner(true);
```
I wasn't sure whether its supported yet , but I will revoke the changes for now.

pdabre12 commented:
Test cases document link: https://ibm.box.com/s/dyhiq2re1izs90wtikxh5psebhzok0a7

pramodsatya reviewed (approved):
Thanks for adding the tests, lgtm.

aditi-pandit commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueryFramework.java:
```diff
@@ -339,6 +341,50 @@ protected void assertTableColumnNames(String tableName, String... columnNames)
+        assertFalse(getQueryRunner().tableExists(session, table));
+    }
+
+    protected void assertCreateTableAsSelectWithoutDrop(String table, @Language("SQL") String query, @Language("SQL") String rowCountQuery)
```
It might be better to add an optional boolean parameter to drop the table in assertCreateTableAsSelect instead of adding all these new variations.

pdabre12 commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java:
```diff
@@ -371,35 +371,6 @@ public void testExplainAnalyzeDDL()
         computeActual("EXPLAIN ANALYZE DROP TABLE orders");
     }
 
-    private void assertExplainAnalyze(@Language("SQL") String query)
```
To minimize code duplication as these functions are used in different tests files,  I moved them to 
`presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueryFramework.java`

pdabre12 commented on presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedTestBase.java:
```diff
@@ -220,11 +218,4 @@ public void testStringFilters()
         assertQuery("SELECT count(*) FROM test_varcharn_filter WHERE shipmode = 'NONEXIST'", "VALUES (0)");
     }
-
-    private void assertExplainAnalyze(@Language("SQL") String query)
```
I moved the function to `presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueryFramework.java`, hence removed it from here.

pdabre12 commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeCtasQueries.java:
```diff
@@ -0,0 +1,104 @@
+    @Override
+    protected void createTables()
+    {
+        QueryRunner queryRunner = (QueryRunner) getExpectedQueryRunner();
```
Will do so in a separate PR after this.

czentgr replied:
Isn't this is the Java query runner (not the native one) to set up the tables from the tpch and create some data to be later used as input for the CTAS of the native query runner?

pdabre12 replied:
Yes, my bad, you are right , it is the Java Runner .

majetideepak replied:
Should we not be testing the create table part with the Native Query RUnner?
What is the purpose of this test?

majetideepak commented:
@pdabre12 can we first refactor the functions and then add CTAS in a separate PR? It is hard to review otherwise.
I also see some commented code under `assertExplainAnalyze`.

pdabre12 commented:
@majetideepak Please take a look.

majetideepak commented:
@pdabre12 I want to avoid refactoring existing API. Can we extend the native runner from AbstractTestDistributedQueries to get additional API?

pdabre12 commented:
> @pdabre12 I want to avoid refactoring the existing API. Can we extend the AbstractTestNativeCtasQueries from AbstractTestDistributedQueries to get the additional API?

Thanks for the review, addressed the comment. 

The one reason I did not want to extend the `AbstractTestNativeCtasQueries` from `AbstractTestDistributedQueries` is while running test cases for a child class,  the test cases in the parent class will be run again.

pdabre12 commented:
This PR is dependent on https://github.com/prestodb/presto/pull/21811.

majetideepak commented:
@pdabre12 can you share how much time these new tests take? It is harder to find this from the CI job.

pdabre12 commented:
> @pdabre12 can you share how much time these new tests take? It is harder to find this from the CI job.

It takes approx ~43 seconds for the new tests to run.

majetideepak commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeCtasQueries.java:
```diff
@@ -0,0 +1,115 @@
+    @Test
+    public void testCreateTableAsSelect()
+    {
+        assertUpdate("CREATE TABLE IF NOT EXISTS test_ctas AS SELECT name, regionkey FROM nation", "SELECT count(*) FROM nation");
```
Should we test `CREATE TABLE` here without `IF NOT EXISTS`?

pdabre12 replied:
I think it might be a good idea to test `CREATE TABLE` without `IF NOT EXISTS` here.
In the other test cases except this, we are testing the `CREATE TABLE` without `IF NOT EXISTS` using the `assertCreateTableAsSelect()` function defined here: https://github.com/prestodb/presto/blob/3df58d1e0f801a331aa9e461375c1027ba5f866f/presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueryFramework.java#L378

majetideepak commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeCtasQueries.java:
```diff
@@ -0,0 +1,115 @@
+        assertTableColumnNames("test_ctas", "name", "regionkey");
+        assertUpdate("DROP TABLE test_ctas");
+        //Some connectors support CREATE TABLE AS but not the ordinary CREATE TABLE. Let's test CTAS IF NOT EXISTS with a table that is guaranteed to exist.
+        assertUpdate("CREATE TABLE IF NOT EXISTS nation AS SELECT orderkey, discount FROM lineitem", 0);
```
I think we can skip this test. We are not testing any Prestissimo path here.

majetideepak commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeCtasQueries.java:
```diff
@@ -0,0 +1,115 @@
+        assertUpdate("DROP TABLE test_ctas");
+        //Some connectors support CREATE TABLE AS but not the ordinary CREATE TABLE. Let's test CTAS IF NOT EXISTS with a table that is guaranteed to exist.
+        assertUpdate("CREATE TABLE IF NOT EXISTS nation AS SELECT orderkey, discount FROM lineitem", 0);
+        assertTableColumnNames("nation", "nationkey", "name", "regionkey", "comment");
```
Remove this as well.

majetideepak commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeCtasQueries.java:
```diff
@@ -0,0 +1,115 @@
+
+        assertCreateTableAsSelect(
+                "test_with_data",
+                "SELECT * FROM orders WITH DATA",
```
This is again a coordinator path as this is the default. Let's remove this test.

majetideepak commented on presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeCtasQueries.java:
```diff
@@ -0,0 +1,115 @@
+                "SELECT count(*) FROM orders");
+
+        assertCreateTableAsSelect(
+                Session.builder(getSession()).setSystemProperty("redistribute_writes", "true").build(),
```
Can you add a comment on what `redistribute_writes` does? What is the default value?

majetideepak reviewed (commented):
@pdabre12 some comments. Thanks.

pdabre12 commented:
@majetideepak Addressed the comments, PTAL.
Thanks.
